### PR TITLE
Update flash_attention kernel from 2.3.6 to 2.5.6

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
Updates FlashAttention kernel code from tag [2.3.6](https://github.com/Dao-AILab/flash-attention/releases/tag/v2.3.6) to [2.5.6](https://github.com/Dao-AILab/flash-attention/releases/tag/v2.5.6).

The usual changes were then re-rellod on top of the modified kernel, changing how dropout saved for backward, removing the head_dim_pad since this would make the kernel inplace mutate and that has a bad interaction with functionalization.

cc @ptrblck @albanD

### CODE REVIEW

**Comment by drisspg** [comment]
Built locally with this cutlass v.3.4.0 and https://github.com/NVIDIA/cutlass/pull/1328 patch

**Comment by drisspg** [comment]
@xinyazhang I am less familiar with the HIP build process and why the build is failing, the signatures of the cuda variants have changed (not the aten funcs) what update needs to be made?

**Comment by jeffdaily** [comment]
@xinyazhang and/or @groenenboomj can you assist with the ROCm failures?

**Comment by drisspg** [comment]
@Xinya Zhang
Thanks!

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu [hunk 1]
```diff
@@ -1,3 +1,4 @@
+#include <string_view>
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/cuda/attention_backward.cu:1]: nit: Why not put it above `type_traits`?

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu [hunk 2]
```diff
@@ -74,6 +74,21 @@ std::tuple<Tensor, Tensor, Tensor> _flash_attention_backward(
...
 
+  // Currently unused args:
+  c10::optional<at::Tensor> alibi_slopes{c10::nullopt};
+
+  bool determinisitic{false};
+  auto& ctx = at::globalContext();
+  if (ctx.deterministicAlgorithms()) {
+    if (ctx.deterministicAlgorithmsWarnOnly()) {
+      TORCH_WARN_ONCE(
+          "Flash Attention defaults to a non-deterministic algorithm. ",
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/cuda/attention_backward.cu:85]: Is this new in 2.5.5?

File: aten/src/ATen/native/transformers/cuda/sdp_utils.cpp [hunk 3]
```diff
@@ -252,11 +252,19 @@ bool check_requires_grad_and_head_dim_gt192_and_sm_ge86_lt90(
   auto dprops = at::cuda::getCurrentDeviceProperties();
   bool is_sm86_or_sm89 = check_sm_version<sm86, sm89>(dprops);
   bool is_head_dim_gt192 = params.query.sym_size(-1) > 192;
-  if (input_requires_grad(params) && is_sm86_or_sm89 && is_head_dim_gt192) {
+  bool is_head_dim_lte224 = params.query.sym_size(-1) <= 224;
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:255]: so now it's `gt192_lte224`?

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu (hunk 1 above)
Comment (drisspg) [reply aten/src/ATen/native/transformers/cuda/attention_backward.cu:1]: can be removed entirely actually

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu (hunk 2 above)
Comment (drisspg) [reply aten/src/ATen/native/transformers/cuda/attention_backward.cu:85]: Yup it is

File: aten/src/ATen/native/transformers/cuda/sdp_utils.cpp (hunk 3 above)
Comment (drisspg) [reply aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:255]: yeah talked with Tri about this, the condition is interesting:
but essentially any input in (192,224] will error because this requires 7 warps which is not divisible by 4

For things greater than 224 we round up to 256 and this is 8 warps, but if you use dropout you use too much shmem so can't run

**Review by cpuhrsch — APPROVED** [review]

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu (hunk 1 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/transformers/cuda/attention_backward.cu:1]: Do you want to do that before landing?
